spec/yaml/swiftnav/sbp/signing.yaml

```diff
@@ -0,0 +1,55 @@
+# Copyright (C) 2018-2022 Swift Navigation Inc.
+# Contact: https://support.swiftnav.com
+#
+# This source is subject to the license found in the file 'LICENSE' which must
+# be be distributed together with this source. All other rights reserved.
+#
+# THIS CODE AND INFORMATION IS PROVIDED "AS IS" WITHOUT WARRANTY OF ANY KIND,
+# EITHER EXPRESSED OR IMPLIED, INCLUDING BUT NOT LIMITED TO THE IMPLIED
+# WARRANTIES OF MERCHANTABILITY AND/OR FITNESS FOR A PARTICULAR PURPOSE.
+
+package: swiftnav.sbp.signing
+description: Messages relating to signatures
+stable: False
+public: False
+include:
+  - types.yaml
+  - gnss.yaml
+definitions:
+  - MSG_ED25519_SIGNATURE:
+      id: 0x0BD2
+      short_desc: ED25519 signature for groups of RTCM messages
+      fields:
+        - signature:
+            type: array
+            fill: u8
+            size: 64
+            desc: ED25519 signature for messages.
+        - fingerprint:
+            type: array
+            fill: u8
+            size: 20
+            desc: SHA-1 fingerprint of the associated certificate.
+        - signed_messages:
+            type: array
```

Comment:
I think we should split this into two messages so that we can provide the client with a message that will allow them to size a buffer that's needed for signing and to easily identify when a group of messages is complete (potentially without having to consult something like a hash map).

So first message would contain something like:
- Count of message and/or size of the signature buffer
- List of CRCs included in signature
- Some sort of signature ID (random or possibly a hash of the CRCs or the signature buffer itself).

Second message would contain:
- Signature ID
- Fingerprint of cert used to generate signature
- The signature itself

Reply:
Isn't this only helpful/useful if we can assume the message stream is well ordered? Or have I misunderstood?

If we can rely on the message stream being ordered then I think we could achieve a similar thing by sending the signature message prior to the messages that it signs, avoiding the need for two messages.

I was of the understanding that we shouldn't/couldn't rely on ordering for RTCM/SBP (even if we might be able to say its OK for the particular customer that this is for).

Reply:
Yeah, getting the feeling I didn't think this through all the way.  There was a question about latency introduced and buffer sizes from the client and was hoping we could come up with a way to size the buffer before the messages arrive while allowing us to pass through the messages.  Obviously if we sent the signature ahead of the messages we could do this easily but it'd mean we'd need to hold these messages in a buffer, generate the signature, then send the signature message ahead of the message group.

One way to do this might be to assume a signature buffer with a fixed maximum size (that is, it'd always be sized to hold 44 messages).  Each block in the buffer would be 255 byte blocks that would be filled by messages.

To find where a message would slot into the buffer we could use the CRC has a hash value, then use open indexing with linear probing to find where to slot the message: https://www.geeksforgeeks.org/implementing-hash-table-open-addressing-linear-probing-cpp/.

This means we'd always be signing a buffer with a fixed size, and many of the bytes would be zero in the signature buffer for smaller groups.

This would require always having 11220 (44*255) bytes for the sig buffer, but this is the upper maximum that would be needed anyway.

Reply:
Interesting approach! 

> 11220 (44*255) bytes for the sig buffer,

We need to sign RTCM messages, not SBP messages, so I think this would actually be 44KB (44*1024).

I also wonder if it is worth building this level of complexity into the protocol? I think a client could implement a similar scheme themselves, if they wanted to, by having a similar fixed-size hashmap (which, if I'm understanding correctly, is essentially what we'd be building server side in your proposal)? In this case, they'd implement something like (excuse the very rough sketch):

```
void on_msg(uint32_t crc, void* buf) {
     fixed_hashmap->insert(crc, buf);
}

bool on_signature(uint32_t* crcs, uint32_t* sig) {
     uint8_t buf[max_size];
     for (crc in crcs) {
       buf.insert(fixed_hashmap->get(crc));
     }

     return verify_sig(sig, buf);
}
```

And in this way, it doesn't preclude clients that are happy to use dynamic memory from implementing the verification in a simpler fashion, if they chose to (by using a standard hashmap).

And not to reroute to start talking about message ordering again, but if we need to account for to non-guaranteed message ordering then I think a client could still need to buffer messages from other groups if the full group hasn't arrived yet. I guess in practice, if a client needed to use fixed buffers, you'd have to start dropping the oldest (by arrival time?) messages once you've buffered as much as possible.

Reply:
> Obviously if we sent the signature ahead of the messages we could do this easily but it'd mean we'd need to hold these messages in a buffer, generate the signature, then send the signature message ahead of the message group.

Also not sure if I see the problem with this. We need to hold messages in a buffer to generate the signature, regardless and the current proposal is to sign messages that are generated together, together.

If latency is the issue, I think this would practicably only be an issue if we started grouping messages that are generated at different times together, which I don't think we're currently intending to do (or at least isn't how I've implemented this for now).

Reply:
> > Obviously if we sent the signature ahead of the messages we could do this easily but it'd mean we'd need to hold these messages in a buffer, generate the signature, then send the signature message ahead of the message group.
> 
> 
> 
> Also not sure if I see the problem with this. We need to hold messages in a buffer to generate the signature, regardless and the current proposal is to sign messages that are generated together, together.
> 
> 
> 
> If latency is the issue, I think this would practicably only be an issue if we started grouping messages that are generated at different times together, which I don't think we're currently intending to do (or at least isn't how I've implemented this for now). (Added you to the go code PR in case it helps get a sense of how this is working ATM)

Yeah, it's maybe a moot point, but I was thinking this approach would essentially allow us to call `send(msg)` on messages as we're putting them into a buffer for a signature, the client would them be able to build its signature buffer in lock step with the server with this approach (and would be able to validate the signature with this "prefilled" buffer as soon as we send it).  But again, the latency improvements here are probably really small.

Reply:
Given there's not likely to be a big improvement in latency for the amount of complexity this brings, I think I'll just leave it as it is. Sending the signature first I think should allow clients to have a simpler implementation. (Though I think there is likely some complexity in properly handling edge cases like messages in a group not arriving - but I don't think there's any nice way around that).

Reply:
Sounds good